File: OmniScript/cs/OmniScript/GraphTemplate.cs
// =============================================================================
// <copyright file="Adapter.cs" company="LiveAction, Inc.">
//  Copyright (c) 2018-2021 Savvius, Inc. All rights reserved.
//  Copyright (c) 2015-2018 Savvius, Inc. All rights reserved.
// </copyright>
// =============================================================================

namespace Savvius.Omni.OmniScript
{
    using System;
    using System.Collections.Generic;
    using System.Xml.Linq;

    public class GraphTemplate
    {
        public const String ClassName = "graph-template";
        public const String RootName = "simplegraph";

        public OmniLogger Logger { get; set; }

        /// <summary>
        /// Gets or sets the description of the graph.
        /// </summary>
        public String Description { get; set; }

        /// <summary>
        /// Gets or sets the Id of the graph.
        /// </summary>
        public OmniId GraphId { get; set; }

        /// <summary>
        /// Gets or sets the Id of the graph.
        /// </summary>
        public OmniId Id { get; set; }

        /// <summary>
        /// Gets or sets the name of the graph.
        /// </summary>
        public String Name { get; set; }

        /// <summary>
        /// Gets or sets the start of the graph.
        /// </summary>
        public uint Start { get; set; }

        /// <summary>
        /// Gets or sets the sample interval of the graph.
        /// </summary>
        public uint SampleInterval { get; set; }

        /// <summary>
        /// Gets or sets the sample count of the graph.
        /// </summary>
        public uint SampleCount { get; set; }

        public GraphTemplate(OmniLogger logger, XElement node=null)
        {
            this.Logger = logger;
            this.Load(node);
        }

        public void Load(XElement node)
        {
            if ((node == null) || (node.Name != ClassName)) return;
            XElement graph = node.Element(RootName);
            if (graph == null) return;
            XElement details = graph.Element("details");
            if (details == null) return;
            IEnumerable<XAttribute> attribs = details.Attributes();
            foreach (XAttribute attrib in attribs)
            {
                String name = attrib.Name.ToString();
                switch (name)
                {
                    case "desc":
                        this.Description = attrib.Value;
                        break;

                    case "id":
                        this.GraphId = new OmniId(attrib.Value);
                        break;

                    case "sample-count":
                        this.SampleCount = Convert.ToUInt32(attrib.Value);
                        break;

                    case "sample-int":
                        this.SampleInterval = Convert.ToUInt32(attrib.Value);
                        break;

                    case "templateid":
                        this.Id = new OmniId(attrib.Value);
                        break;

                    case "title":
                        this.Name = attrib.Value;
                        break;

                    case "start":
                        this.Start = Convert.ToUInt32(attrib.Value);
                        break;
                }
            }
        }
    }
}

File: OmniScript/py/OmniScript/omniscript/graphtemplate.py
"""GraphTemplate class.
"""
#Copyright (c) WildPackets, Inc. 2013-2015. All rights reserved.

import os
import sys
import logging
import xml.etree.ElementTree as ET

import omniscript

from omniid import OmniId


find_attribs = ['name', 'id']


class GraphTemplate(object):
    """The GraphTemplate class has the attributes of an Graph Template.
    The 
    :func:`get_graph_template_list() 
    <omniscript.omniengine.OmniEngine.get_graph_template_list>`
    function returns a list of GraphTemplate objects.
    """

    id = OmniId()
    """The graph template's identifier."""

    name = ''
    """The name of the graph template."""

    description = ''
    """The description of the graph template."""

    graph_id = OmniId()
    """The graph identifier of the graph template."""

    start = 0
    """The start value"""

    sample_interval = 0
    """The sample interval"""

    sample_count = 0
    """The sample count"""

    def __init__(self, criteria):
        self.id = OmniId(True) if criteria is None else GraphTemplate.id
        self.name = GraphTemplate.name
        self.description = GraphTemplate.description
        self.graph_id = OmniId(True) if criteria is None else GraphTemplate.graph_id
        self.start = GraphTemplate.start
        self.sample_interval = GraphTemplate.sample_interval
        self.sample_count = GraphTemplate.sample_count
        if isinstance(criteria, basestring):
            xml = ET.parse(criteria)
            self._load(xml)
        elif isinstance(criteria, ET.Element):
            self._load(criteria)

    def __repr__(self):
        return 'GraphTemplate: %s' % self.name

    def _load(self, element):
        graph = element.find('simplegraph')
        if graph is not None:
          details = graph.find('details')
          if details is not None:
            for attrib in details.items():
                if attrib[0] == 'templateid':
                    self.id = OmniId(attrib[1])
                elif attrib[0] == 'title':
                    self.name = attrib[1]
                elif attrib[0] == 'desc':
                    self.description = attrib[1]
                elif attrib[0] == 'id':
                    self.graph_id =  OmniId(attrib[1])
                elif attrib[0] == 'start':
                    self.start = int(attrib[1])
                elif attrib[0] == 'sample-int':
                    self.sample_interval = int(attrib[1])
                elif attrib[0] == 'sample-count':
                    self.sample_count = int(attrib[1])
            if self.id is None:
                self.id = OmniId(None)


def _create_graph_template_list(xml_list):
    lst = []
    graphs = omniscript._parse_command_response(xml_list, 'graph-collection')
    if graphs is not None:
        for template in graphs.findall('graph-template'):
            lst.append(GraphTemplate(criteria=template))
    lst.sort(key=lambda x: x.name)
    return lst


def find_graph_template(graphs, value, attrib=find_attribs[0]):
    """Finds a graph_template in the list"""
    if (not graphs) or (attrib not in find_attribs):
        return []

    if len(graphs) == 0:
        return []

    if isinstance(value, GraphTemplate):
        _value = value.id
        attrib = 'id'
    else:
        _value = value

    return next((i for i in graphs if getattr(i, attrib) == _value), [])
